ci/functional/job_cleanup.sh

```diff
@@ -17,15 +17,18 @@ rm -rf install/lib/daos/TESTING/ftest/avocado/job-results/job-*/html/
 
 # Remove the latest avocado symlink directory to avoid inclusion in the
 # jenkins build artifacts
-if [ -e install/lib/daos/TESTING/ftest/avocado/job-results/latest ]; then
-  unlink install/lib/daos/TESTING/ftest/avocado/job-results/latest
-fi
+DIRS=(install/lib/daos/TESTING/ftest/avocado/job-results/*)
+for results_dir in "${DIRS[@]}"; do
+    if [[ -L "${results_dir}" && -d "${results_dir}" ]]; then
+      unlink "${results_dir}"
+    fi
+done
```

Comment:
> Looking at the Functional HW Small output, it appears that the `-L` test passes and the `-d` test failed for `install/lib/daos/TESTING/ftest/avocado/job-results/latest`

Yes, of course.  `-L` and `-d` are mutually exclusive conditions.  They can **never** be true.

> What's interesting is `install/lib/daos/TESTING/ftest/avocado/job-results/latest` does appear in the `ls -d` output:
> 
> ```
> [2021-02-06T04:52:43.322Z] ++ ls './*daos.log*' './*daos_agent.log*'
> [2021-02-06T04:52:43.322Z] + arts=
> [2021-02-06T04:52:43.322Z] ++ ls -d install/lib/daos/TESTING/ftest/avocado/job-results/aggregation install/lib/daos/TESTING/ftest/avocado/job-results/control install/lib/daos/TESTING/ftest/avocado/job-results/daos_vol install/lib/daos/TESTING/ftest/avocado/job-results/datamover install/lib/daos/TESTING/ftest/avocado/job-results/io install/lib/daos/TESTING/ftest/avocado/job-results/latest install/lib/daos/TESTING/ftest/avocado/job-results/network
> [2021-02-06T04:52:43.583Z] + arts='install/lib/daos/TESTING/ftest/avocado/job-results/aggregation
> [2021-02-06T04:52:43.583Z] install/lib/daos/TESTING/ftest/avocado/job-results/control
> [2021-02-06T04:52:43.583Z] install/lib/daos/TESTING/ftest/avocado/job-results/daos_vol
> [2021-02-06T04:52:43.583Z] install/lib/daos/TESTING/ftest/avocado/job-results/datamover
> [2021-02-06T04:52:43.583Z] install/lib/daos/TESTING/ftest/avocado/job-results/io
> [2021-02-06T04:52:43.583Z] install/lib/daos/TESTING/ftest/avocado/job-results/latest
> [2021-02-06T04:52:43.583Z] install/lib/daos/TESTING/ftest/avocado/job-results/network'
> ```

Indeed!  My observation also.  I added more debug even:

```
[2021-02-09T10:22:29.960Z] + '[' -e install/lib/daos/TESTING/ftest/avocado/job-results/latest ']'
[2021-02-09T10:22:29.960Z] + pwd
[2021-02-09T10:22:29.960Z] /data/jenkins/jenkins-1/ci_vm_wolf-102vm1-9/workspace/daos-stack_daos_PR-4470
[2021-02-09T10:22:29.960Z] + ls -l install/lib/daos/TESTING/ftest/avocado/job-results
[2021-02-09T10:22:29.960Z] total 24
[2021-02-09T10:22:29.960Z] drwxrwxr-x 15 lcl_builder lcl_builder 4096 Feb  9 10:22 cart
[2021-02-09T10:22:29.960Z] drwxrwxr-x  3 lcl_builder lcl_builder 4096 Feb  9 09:54 container
[2021-02-09T10:22:29.960Z] drwxrwxr-x  4 lcl_builder lcl_builder 4096 Feb  9 10:10 io
[2021-02-09T10:22:29.960Z] lrwxrwxrwx  1 lcl_builder lcl_builder   55 Feb  9 10:22 latest -> /var/tmp/ftest/avocado/job-results/./cart/group_test.py
[2021-02-09T10:22:29.960Z] drwxrwxr-x  3 lcl_builder lcl_builder 4096 Feb  9 10:07 network
[2021-02-09T10:22:29.960Z] drwxrwxr-x  4 lcl_builder lcl_builder 4096 Feb  9 10:18 pool
[2021-02-09T10:22:29.960Z] drwxrwxr-x  3 lcl_builder lcl_builder 4096 Feb  9 10:01 security
```
 
> Now because this stage is testing with the RPMs it does run
> 
> https://github.com/daos-stack/daos/blob/f451fb49aa68256ca1ae37dc5bd3134b5ba97b8f/ci/functional/job_cleanup.sh#L7-L14
> 
> . Maybe we should exclude the `latest` directory as part of the
> https://github.com/daos-stack/daos/blob/f451fb49aa68256ca1ae37dc5bd3134b5ba97b8f/ci/functional/job_cleanup.sh#L12-L13
> 
> command.

Maybe.  But none of that answers this strange behaviour that we are seeing.  We really should understand why things are not working as they very clearly seem they should be.  There could be an issue lurking there that is going to have other ripple effects.

I have added some more debug to my test that produced the above debug output.

Reply:
Now that we understand why this happens (thanks Brian) I think it can be marked resolved as the exclusion of the "latest" directory in the Jenkins artifacts has been resolved by [excluding it from the move](https://github.com/daos-stack/daos/blob/ae26b4eceb97cb47635fab2cba0d2ffea1edf401/ci/functional/job_cleanup.sh#L24-L30).